feat(blaze): Add visual error indicator for development

wreiske opened the pull request:
- Add error badge in bottom-left corner showing error count
- Add modal with detailed error information and stack traces
- Implement graceful failure with inline placeholders for missing templates
- Add error deduplication to prevent duplicate entries
- Integrate with HMR to auto-clear errors when templates are fixed
- Add accessibility features (ARIA labels, keyboard navigation)
- Auto-disable in production mode
- Add TypeScript definitions for new public API

## Error Recovery

### Runtime errors (missing templates)
`Spacebars.include` now catches errors from `templateOrFunction()` (e.g. `No such template: broken`) instead of letting them crash the entire page render. The error is reported to the error indicator, and an inline placeholder is rendered in place of the missing template. Other templates continue to render normally.

### Compile-time errors (invalid template syntax)
Spacebars parse errors (e.g. using `{{#if}}` blocks inside HTML attributes) no longer crash the Meteor build with "Errors prevented startup". Instead, the broken template is compiled into a fallback that renders an inline error placeholder with full details (file, line, error message). The error is also reported to `Blaze._errorIndicator` at runtime. A `console.warn` is emitted during build so the error remains visible in the server terminal.

BREAKING CHANGE: Missing template errors now show placeholders instead of throwing, which may change error handling behavior in some edge cases.

## Before Change

Before this PR, blaze template issues would just result in the entire app not working.

<img width="919" height="645" alt="Screenshot 2025-11-25 at 1 23 27 AM" src="https://github.com/user-attachments/assets/f53f9ec4-510f-4913-866d-52b78018968e"/>

## After Change

Now the app works, AND it will gracefully continue rendering the rest of the app! Also a nice little error window to know where you messed up.

<img width="943" height="751" alt="Screenshot 2025-11-25 at 1 22 11 AM" src="https://github.com/user-attachments/assets/c75ccd66-ca30-4d00-aeca-323f9c2820e1"/>
<img width="943" height="751" alt="Screenshot 2025-11-25 at 1 22 22 AM" src="https://github.com/user-attachments/assets/077971df-75b8-40d5-8f4c-88401470c505"/>

wreiske committed:
feat(blaze): Add visual error indicator for development

- Add error badge in bottom-left corner showing error count
- Add modal with detailed error information and stack traces
- Implement graceful failure with inline placeholders for missing templates
- Add error deduplication to prevent duplicate entries
- Integrate with HMR to auto-clear errors when templates are fixed
- Add accessibility features (ARIA labels, keyboard navigation)
- Auto-disable in production mode
- Add TypeScript definitions for new public API

BREAKING CHANGE: Missing template errors now show placeholders instead of
throwing, which may change error handling behavior in some edge cases.

wreiske committed:
Improve error reporting in HTML compiler and bump versions

Enhances Blaze template compilation error handling to display errors in the browser overlay and console, making them more visible to developers. Also fixes an instanceof check in html-scanner-tests.js and ensures parse errors are properly propagated in html-tools. Updates package versions and interdependencies to 2.0.1 for consistency.

jankapunkt committed:
Merge branch 'master' into feature/error-indicator

wreiske committed:
Merge branch 'master' into feature/error-indicator

jankapunkt committed:
Merge branch 'release-3.1.0' into feature/error-indicator

jankapunkt commented:
@wreiske I changed the merge base from `master` to `release-3.1.0` so we can ship this with the upcoming release. May you take a look at this again? The CI timeout should be fixed by now.

jankapunkt committed:
Merge branch 'release-3.1.0' into feature/error-indicator

wreiske committed:
Merge branch 'release-3.1.0' into feature/error-indicator

wreiske committed:
fix: preserve original throw behavior in lookup.js

Address review feedback from @italojs to avoid breaking changes:

- Restore original throw behavior for missing templates, functions,
  directives, and non-function calls in lookup.js
- Remove explicit _reportException calls from lookup.js (errors are
  already caught by _wrapCatchingExceptions in the rendering pipeline
  and reported via _reportException, which feeds the error indicator)
- Restore original caching-html-compiler compile error handling
  (return null on CompileError) to prevent build system conflicts

The error indicator still works through two channels:
1. _reportException in exceptions.js feeds caught errors to the indicator
2. window.addEventListener('error') in errorIndicator.js catches uncaught
   Blaze-related errors

All 416 tests pass.

wreiske committed:
fix(spacebars): Gracefully handle missing template errors in Spacebars.include

Wrap templateOrFunction() in a try-catch inside the Spacebars.include
autorun so that errors like 'No such template: broken' no longer crash
the entire page render. Instead, the error is stored in the reactive
var, reported via Blaze._reportException (which feeds the error
indicator), and an inline error placeholder is rendered in place of the
missing template.

wreiske committed:
fix(templating-tools): Gracefully handle Spacebars compile errors instead of crashing the build

When a template has a Spacebars parse error (e.g. using {{#if}} blocks
inside HTML attributes), the build now continues instead of failing with
'Errors prevented startup'. The broken template is replaced with a
fallback that renders an inline error placeholder showing the compile
error details, and the error is reported to Blaze._errorIndicator at
runtime. Other templates in the same file compile normally.

A console.warn is emitted during build so the error is still visible in
the server terminal.

wreiske committed:
fix(blaze): Gracefully handle render errors in view autorun

Wrap view._render() in a try-catch inside the doRender autorun so that
runtime errors (e.g. 'No such function: test' from missing helpers)
render an inline error placeholder instead of crashing the entire page.
The error is reported via Blaze._reportException which feeds the error
indicator.

wreiske committed:
fix: gate graceful error fallback to development only, catch binding errors

- compile-tags-with-spacebars: Only generate inline error placeholders
  when HMR is available (development). During meteor build, compilation
  errors now properly throw CompileError and fail the build as expected.

- builtins: Wrap _createBinding autorun in try/catch so runtime errors
  (e.g. 'No such function: test') are caught and reported via
  Blaze._reportException instead of crashing as uncaught exceptions.
  This routes them to the error indicator during development.

wreiske committed:
refactor: DRY up error placeholders, fix memory leak, fix test failures

- Replace unused Blaze._errorPlaceholder with shared _renderErrorPlaceholder helper
- Extract _ERROR_PLACEHOLDER_STYLE constant used by view.js, spacebars-runtime.js,
  and compile-tags-with-spacebars.js (was duplicated 3x with inconsistencies)
- Fix memory leak in errorIndicator.js: destroy() now properly removes
  window 'error' and 'unhandledrejection' listeners
- Fix clearAll duplication: clear button handler now calls clearAll() directly
- Revert overly-broad try/catch in view.js and builtins.js that swallowed
  errors the existing Blaze error pipeline already handles
- Add _throwNextException check in spacebars-runtime.js catch block so
  test assertions continue to work
- Update 'unfound template' tests to verify graceful degradation behavior

All 416 tests pass.

wreiske committed:
fix: restore try/catch in builtins.js _createBinding with _throwNextException guard

Binding errors (e.g. calling a non-existent function in {{#if test ...}})
were crashing the entire page render. Restore the try/catch but add
the _throwNextException check so tests still work.

All 416 tests pass.

jankapunkt commented:
If anyone else wants to test and review this PR you need to checkout via

```shell
git fetch origin pull/481/head:feature/visual-error-indicator
git checkout feature/visual-error-indicator
```

jankapunkt committed:
Merge branch 'release-3.1.0' into feature/error-indicator

jankapunkt commented on packages/templating-tools/package.js:
```diff
@@ -1,7 +1,7 @@
   name: 'templating-tools',
   summary: "Tools to scan HTML and compile tags when building a templating package",
-  version: '2.0.0',
+  version: '2.0.1',
```
can you please reverse the version numbers? we will bump them on the alpha, beta, rc etc. for 3.1.0

jankapunkt commented on packages/blaze/package.js:
```diff
@@ -1,7 +1,7 @@
   name: 'blaze',
   summary: "Meteor Reactive Templating library",
-  version: '3.0.2',
+  version: '3.1.0',
```
can you please reverse the version numbers? we will bump them on the alpha, beta, rc etc. for 3.1.0

jankapunkt commented on packages/caching-html-compiler/package.js:
```diff
@@ -2,7 +2,7 @@
   name: 'caching-html-compiler',
   summary: 'Pluggable class for compiling HTML into templates',
-  version: '2.0.0',
+  version: '2.0.1',
```
can you please reverse the version numbers? we will bump them on the alpha, beta, rc etc. for 3.1.0

jankapunkt commented on packages/html-tools/package.js:
```diff
@@ -1,7 +1,7 @@
   name: 'html-tools',
   summary: "Standards-compliant HTML tools",
-  version: '2.0.0',
+  version: '2.0.1',
```
can you please reverse the version numbers? we will bump them on the alpha, beta, rc etc. for 3.1.0

jankapunkt commented on packages/spacebars-compiler/package.js:
```diff
@@ -1,15 +1,15 @@
 
   api.use('htmljs@2.0.1');
-  api.use('html-tools@2.0.0');
+  api.use('html-tools@2.0.1');
```
can you please reverse the version numbers? we will bump them on the alpha, beta, rc etc. for 3.1.0

jankapunkt commented on packages/spacebars-compiler/package.js:
```diff
@@ -1,15 +1,15 @@
   name: 'spacebars-compiler',
   summary: "Compiler for Spacebars template language",
-  version: '2.0.0',
+  version: '2.0.1',
```
can you please reverse the version numbers? we will bump them on the alpha, beta, rc etc. for 3.1.0

jankapunkt commented:
@wreiske if you are busy I can also revert the versions if you want

wreiske committed:
fix: revert version number bumps per maintainer request

Version bumps will be handled during the 3.1.0 alpha/beta/RC release process.

wreiske commented:
Done — reverted all version number bumps in 15c3b0c. The five `package.js` files (`blaze`, `caching-html-compiler`, `html-tools`, `spacebars-compiler`, `templating-tools`) are back to their original versions, including the `api.use()` dependency references.

jankapunkt commented:
@wreiske I tested this using one of our medium sized apps that heavily uses dynamically imported templates and things work as expected except when adding faulty code in HMR mode:

When adding `<button {{#if counter}}disabled{{/if}}>Click Me</button>` to an already rendered template, I get no visual error indicators, but only the (server) console message 

```
=> Client modified -- refreshingWarning: Compile error in imports/ui/layout/navSide/navSide.html:14: Reactive HTML attributes must either have a constant name or consist of a single {{helper}} providing a dictionary of names and values.  A template tag of type BLOCKOPEN is not allowed here.
...  </a>       <button {{#if counter}}disab...
                        ^
The app will still run, but the affected template will show an error placeholder.
```

If I instead add something like `{{#if foo moo}}<p>hi</p>{{/if}}` , where `foo` does not exist, there is neither a visual error indicator, nor a console message.

Note, that this happens in a dynamically imported module and HMR active. The other behavior works well, even with dynamic imported modules.

jankapunkt committed:
Merge branch 'release-3.1.0' into feature/error-indicator

jankapunkt commented on packages/blaze/errorIndicator.js:
```diff
@@ -0,0 +1,720 @@
+ * 
+ * @fileoverview Client-side error indicator for Blaze templates
+ */
+
```
So I went through all the code here and the only question I have left is:

If we hard-fail on production builds, then wouldn't this file be obsolete? Meaning - Could we exclude this code if `Meteor.isProduction` to save the few KB for client load optimization?

**EDIT** - I maybe was a bit unspecific - while the code contains already checks for production mode, I think he majority of this file code be excluded at all on production as the other places have guard clauses that check for function existence before calling them.

I know with rspack etc. not an issue but I think we should, for backward compat, use the good ol' code splitting here, right?

Apart from that I would merge this and improve on request, once 3.1.0 alpha is out.
